jsf-ri/src/main/java/com/sun/faces/renderkit/html_basic/AjaxBehaviorRenderer.java

```diff
@@ -340,24 +350,21 @@ private static void appendIds(UIComponent component,
                 id.equals("@form") || id.equals("@this")) {
                 builder.append(id);
             } else {
-                builder.append(getResolvedId(component, id));
+                if (searchExpressionContext == null)  {
+                    searchExpressionContext = SearchExpressionContext.createSearchExpressionContext(
+                            facesContext, component, EXPRESSION_HINTS, null);
+                }
+                
+                builder.append(resolveExpression(searchExpressionContext, id));
             }
         }
 
         builder.append("'");
     }
 
-    // Returns the resolved (client id) for a particular id.
-    private static String getResolvedId(UIComponent component, String id) {
-
-        UIComponent resolvedComponent = component.findComponent(id);
-        if (resolvedComponent == null) {
-            if (id.charAt(0) == UINamingContainer.getSeparatorChar(FacesContext.getCurrentInstance())) {
-                return id.substring(1);
-            }
-            return id;
-        }
-
-        return resolvedComponent.getClientId();
+    // Returns the resolved (client id) for a particular expression.
+    private static String resolveExpression(SearchExpressionContext searchExpressionContext, String expression) {
+        SearchExpressionHandler handler = searchExpressionContext.getFacesContext().getApplication().getSearchExpressionHandler();
```

Comment:
Rename to resolveClientId. Though I wonder how necessary it is to have an oneliner in a private utility method.

For remainder, nice!